proposals/0530-strict-box-denesting.rst

```diff
@@ -0,0 +1,613 @@
+
+   Here, the ``Named`` newtype can be used to somewhat emulate
+   dependent types and write interfaces that provide unusually strong
+   compile-time guarantees.  The ``Proof (p n)`` field of ``WithProof``
+   unpacks to a zero-width field of type ``p n ~# True`` in the
+   constructor worker, so ``WithProof`` satisfies  conditions 1-4.
+   But that ``Proof (p n)`` field carries compile-time evidence
+   that may be necessary to satisfy the typechecker when using an emulated
+   dependently-typed interface, so throwing it away by using a newtype
+   around ``exists (n :: t). t `Named` n`` would be unacceptable.
+
+   ..
+     A wrapper around a more complex type like
+     ``exists (n :: t). (# UnboxedRep (Proof (p n)), t `Named` n #)``
+     is perhaps plausible in some distant future.
+
+
+4. With the ``LinearTypes`` extension, a data constructor can have any
+   multiplicity on its fields, while a newtype constructor must be
+   linear in its field.  Thus, the analogous newtype to this variant on
+   ``linear-base``'s ``Ur`` is not allowed:
+
+   ::
+
+     data StrictUr a where
+       StrictUr :: !a %Many -> Ur a
+
+   (It is not possible to consume a value of type ``StrictUr a``
+   linearly without forcing the contents of its "unrestricted" field,
+   making this type slightly less flexible than the lazy ``Ur``.)
+
+
+5. There may be more than one constructor in the data type.  In this
+   case, condition 4(ii) must apply.  The original motivating
+   example for this condition is ``GHC.ForeignPtr.ForeignPtrContents``:
+
+   ::
+
+      data ForeignPtrContents
+        = PlainForeignPtr !(IORef Finalizers)
+        | FinalPtr
+        | MallocPtr (MutableByteArray# RealWorld) !(IORef Finalizers)
+        | PlainPtr (MutableByteArray# RealWorld)
+
+   This type is used to optionally provide support for finalizers on
+   ``ForeignPtr``\ s, as well as to keep alive GC-managed pinned
+   buffers that may be stored in ``ForeignPtr``\ s.  Depending on a
+   potential re-ordering of its constructors, either ``PlainPtr`` or
+   ``PlainForeignPtr`` (but not both) could satisfy condition 4(ii)
+   and be made denestable.  The former would remove two words of
+   allocation overhead in the buffer creation for any fresh
+   ``StrictByteString``, while the latter would make attaching
+   finalizers to external buffers received from the FFI more
+   efficient.
+
+
+..
+   The links in "many examples of wrapper types" in the next paragraph
+   provide almost exactly the 4(i)-denestable constructors found in the
+   boot libraries of a prototype GHC detecting them, with three omissions.
+   Two are the examples above for reasons 1 and 2: SomeTypeRep and SBS.
+   The last one is base:Control.Concurrent.QSemN.QSemN, which I omitted
+   because it could have been written as a newtype around IORef, but was not,
+   for reasons unknown to me.
+
+All five of these situations are fairly uncommon in general-purpose
+Haskell code.  However,
```

Comment:
it's good to give a lot of links, but in the rendered from it looks as if there are only four, so most readers will miss most of the links. Lay them out